src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py

```diff
@@ -0,0 +1,165 @@
+    "containerFraction": real number with the container fraction to be distributed (TBFD)
+    "replicationGrouping": string with a constant value (DATASET or ALL, to be in sync with Rucio)
+    "activatedOn": int, seconds since epoch in GMT timezone
+    "deactivatedOn": int, seconds since epoch in GMT timezone
+    "active": boolean,
+    "pileupSizeGB": integer, current size of the pileup in GB (integer)
+    "rulesList: list of strings (rules) used to lock the pileup id
+}
+
+The data flow should be done via list of objects, e.g.
+[<pileup object 1>, <pileup object 2>, ..., <pileup object n>]
+"""
+
+# system modules
+import json
+
+# WMCore modules
+from WMCore.MicroService.Tools.Common import getMSLogger
+from WMCore.Lexicon import dataset
+from Utils.Timers import gmtimeSeconds
+
+
+class MSPileupObj(object):
+    """
+    MSPileupObj defines MSPileup data stucture
+    """
+    def __init__(self, pdict, verbose=None, logger=None, validRSEs=None):
+        self.logger = getMSLogger(verbose, logger)
+        if not validRSEs:
+            validRSEs = []
+        self.validRSEs = validRSEs
+
+        self.data = {
+            'pileupName': pdict.get('pileupName', ''),
+            'pileupType': pdict.get('pileupType', 'classic'),
+            'insertTime': pdict.get('insertTime', gmtimeSeconds()),
+            'lastUpdateTime': pdict.get('lastUpdateTime', gmtimeSeconds()),
+            'expectedRSEs': pdict.get('expectedRSEs', []),
+            'currentRSEs': pdict.get('currentRSEs', []),
+            'fullReplicas': pdict.get('fullReplicas', 0),
+            'campaigns': pdict.get('campaigns', []),
+            'containerFraction': pdict.get('containerFraction', 0.0),
+            'replicationGrouping': pdict.get('replicationGrouping', ""),
+            'activatedOn': pdict.get('activatedOn', gmtimeSeconds()),
+            'deactivatedOn': pdict.get('deactivatedOn', gmtimeSeconds()),
+            'active': pdict.get('active', False),
+            'pileupSize': pdict.get('pileupSize', 0),
+            'ruleList': pdict.get('ruleList', [])}
+        if not self.validate(self.data):
+            msg = 'MSPileup input is invalid'
+            raise Exception(msg)
+
+    def __str__(self):
+        """
+        Return human readable representation of pileup data
+        """
+        return json.dumps(self.data, indent=4)
+
+    def getPileupData(self):
+        """
+        Get pileup data
+        """
+        return self.data
+
+    def validate(self, pdict=None):
+        """
+        Validate data according to its schema. If data is not provided via
+        input pdict parameter, the validate method will validate internal
+        data object.
+
+        :param pdict: input data dictionary (optional)
+        :return: boolean, result of validation
+        """
+        if not pdict:
+            pdict = self.data
+        docSchema = self.schema()
+        for key, val in pdict.items():
```

Comment:
Am I correct in saying that if the end user provides extra key/value pairs, that this validate function would not catch them and would let the request proceed? If so, then I think we should ensure that the end-users are providing the correct data structure, not more, not less fields.

Reply:
If user will provide a key:value pair which is not in a schema it will be caught in next line (https://github.com/dmwm/WMCore/pull/11438/files/afa5aa802ee6d41b4110d7d2fe3b8557830adbd0#diff-21c86cd53529a8674a70ffc7953fbbfc5d8ff98acdd7bc6962d257bd55e673dcR92) and dict will not be validated.

Reply:
well, it is tricky to answer. Currently, if we create object and a key is not provided, then object will have default value defined in a schema for that key. But if we validate dictionary which is not created as object, then your observation may be valid. To avoid that I'll add the check for keys.

Reply:
the changes are committed.

Reply:
Valentin, you are right. If we always validate `self.data`, then the parameters will always match what is expected in the schema.

Thinking about a use POSTing a new pileup configuration, I think it would be ideal to validate the user-provided parameters, such that they know in case something is missing/extra/wrong. To deal with such scenario, maybe we would have to implement something like:
"""
pre-validation: validate user-provided pdict (validate only the keys! Disregarding values and data type)
initialize the MSPileupObj with the user-provided data
post-validation: validate self.data (expected object schema, but now validating values and value type).
"""

what do you think? Are we missing anything in here?

Reply:
I would like to avoid over-complicated steps. It should be one validation call, otherwise we'll need to maintain if/else logic in HTTP service APIs. I already thought about it, and if you checked my PR https://github.com/dmwm/WMCore/pull/11443 and look at 
- HTTP post request https://github.com/dmwm/WMCore/pull/11443/files#diff-5741164102ebe9ce2987f9e018708d8a39c8d3b8bc70701fd3446898e9189342R376
- and, `MSPileup` create API, https://github.com/dmwm/WMCore/pull/11443/files#diff-559a1ede5befe7a1ba862bb3a6cbf18fe85dcb9844fadb719f37da82db4e03d9R121
you'll see that it will call only `createPileup` data layer API, which in this PR does the following :
- create MSPileupObj (https://github.com/dmwm/WMCore/pull/11438/files#diff-df26107439b08bc216635653813392ddebb3dc94f0514c164a15d5115d790484R93) which by itself
- calls `validate` API, https://github.com/dmwm/WMCore/pull/11438/files#diff-21c86cd53529a8674a70ffc7953fbbfc5d8ff98acdd7bc6962d257bd55e673dcR63

So, the HTTP post request will call create, the create will creates MSPileupObj which by itself will validate data. Therefore, if validate performs all the steps we do not need more complex logic. I may be wrong, but I think it is sufficient, and I do not think we need to over complicate existing data layer. But, if we'll find it later that it can be useful we can add such logic and split validate into multiple steps.

Reply:
In that case, your latest commit L93-L98 is not really needed, given that `pdict` in this validate method will always be equal to schema.
I would suggest to add a "NOTE" or "TODO" comment in the beginning of this method, saying that not all of the original user-provided payload is validated. Just in case we have problems in the near future and we need to improve it.

Reply:
I think the doc string of `validate` clearly describe the two cases, please see https://github.com/dmwm/WMCore/pull/11438/files#diff-21c86cd53529a8674a70ffc7953fbbfc5d8ff98acdd7bc6962d257bd55e673dcR82 I doubt we need anymore NOTEs or TODOs here, but we do need to get better understanding how we'll create and use objects. So far it is far from clear for me, and therefore it is better to keep the logic as is and rely on docstring which specifically describes the behavior.

Reply:
Okay, that works as well.